### ISSUE TITLE
Cabal coverage feature does not work well with custom build directory

### ISSUE THREAD

**Issue by sergv** [haskell/cabal#2225]
When trying to get hpc for package usngcustom build directory, e.g. `/tmp/dist`, the `hpc markup` tool does not produce the final report because it does not check the right directory for *.mix files.

It seems that `hpc markup` assumes that build directory is located in current directory so it appends "./" to search paths by default (see last line of the log). I can instruct `hpc markup` to look for absolute directories as well as local oness by passing `--hpc-options='--srcdir="" --srcdir="."'` to `cabal configure`. But this produces another unintended error from `hpc sum` that does not have `--srcdir` option, but cabal passes it anyway (for this please consult the second log).

Sample build log that outlines the problem:

```
cd /home/sergey/projects/haskell/projects/sicp/scheme/ && \
cabal configure --enable-library-coverage --disable-library-profiling --disable-executable-profiling --disable-split-objs --builddir /tmp/dist --enable-tests && \
cabal build -v1 --builddir /tmp/dist --ghc-options="-j4 -ferror-spans" && \
cabal test -v2 --builddir /tmp/dist --show-details=always
Resolving dependencies...
Configuring scheme-0.1...
Building scheme-0.1...
Preprocessing library scheme-0.1...
unused terminals: 1
[ 1 of 12] Compiling SexpFrontend.LexTok ( src/SexpFrontend/LexTok.hs, /tmp/dist/build/SexpFrontend/LexTok.o )
[ 2 of 12] Compiling Utils            ( src/Utils.hs, /tmp/dist/build/Utils.o )
[ 3 of 12] Compiling SexpFrontend.Lexer ( /tmp/dist/build/SexpFrontend/Lexer.hs, /tmp/dist/build/SexpFrontend/Lexer.o )
[ 4 of 12] Compiling ALaCarte         ( src/ALaCarte.hs, /tmp/dist/build/ALaCarte.o )

src/ALaCarte.hs:159:1-23: Warning:
    Pattern match(es) are non-exhaustive
    In an equation for ‘remOuterA’: Patterns not matched: _ :&: _

src/ALaCarte.hs:(404,5)-(406,30): Warning:
    Pattern match(es) are non-exhaustive
    In an equation for ‘g’: Patterns not matched: _ :^: _
[ 5 of 12] Compiling ALaCarte.TH      ( src/ALaCarte/TH.hs, /tmp/dist/build/ALaCarte/TH.o )
[ 6 of 12] Compiling Types            ( src/Types.hs, /tmp/dist/build/Types.o )
[ 7 of 12] Compiling Eval.EvalM       ( src/Eval/EvalM.hs, /tmp/dist/build/Eval/EvalM.o )
[ 8 of 12] Compiling Eval.ExtractProgram ( src/Eval/ExtractProgram.hs, /tmp/dist/build/Eval/ExtractProgram.o )

src/Eval/ExtractProgram.hs:(185,8)-(195,59): Warning:
    Pattern match(es) are non-exhaustive
    In an equation for ‘destructLetBinding’:
        Patterns not matched:
            Term (_ :&: _)
            Term ((_ :^: _) :^: _)

src/Eval/ExtractProgram.hs:(205,12)-(219,94): Warning:
    Pattern match(es) are non-exhaustive
    In an equation for ‘go’:
        Patterns not matched:
            (_, _) (Term (_ :&: _))
            (_, _) (Term ((_ :^: _) :^: _))
[ 9 of 12] Compiling Eval.Desugar     ( src/Eval/Desugar.hs, /tmp/dist/build/Eval/Desugar.o )
[10 of 12] Compiling Parse            ( src/Parse.hs, /tmp/dist/build/Parse.o )
[11 of 12] Compiling SexpFrontend.Parser ( /tmp/dist/build/SexpFrontend/Parser.hs, /tmp/dist/build/SexpFrontend/Parser.o )
[12 of 12] Compiling Eval             ( src/Eval.hs, /tmp/dist/build/Eval.o )
In-place registering scheme-0.1...
Preprocessing test suite 'scheme-test' for scheme-0.1...
[1 of 1] Compiling Main             ( test/Main.hs, /tmp/dist/build/scheme-test/scheme-test-tmp/Main.o )
Linking /tmp/dist/build/scheme-test/scheme-test ...
Using a sandbox located at
/home/sergey/projects/haskell/projects/sicp/scheme/.cabal-sandbox
Reading available packages...
Reading installed packages...
/home/sergey/projects/haskell/ghc/local/bin/ghc-pkg dump '--package-db=/home/sergey/projects/haskell/projects/sicp/scheme/.cabal-sandbox/x86_64-linux-ghc-7.8.3-packages.conf.d' -v0
/home/sergey/projects/haskell/ghc/local/bin/ghc --print-libdir
Found no modified add-source deps.
Component build order: library, test suite 'scheme-test'
creating /tmp/dist/build
creating /tmp/dist/build/autogen
Building scheme-0.1...
Preprocessing library scheme-0.1...
Building library...
creating /tmp/dist/build
/home/sergey/projects/haskell/ghc/local/bin/ghc --make -fbuilding-cabal-package -O2 -static -dynamic-too -dynosuf dyn_o -dynhisuf dyn_hi -outputdir /tmp/dist/build -odir /tmp/dist/build -hidir /tmp/dist/build -stubdir /tmp/dist/build -i -i/tmp/dist/build -isrc -i/tmp/dist/build/autogen -I/tmp/dist/build/autogen -I/tmp/dist/build -optP-include -optP/tmp/dist/build/autogen/cabal_macros.h -package-name scheme-0.1 -hide-all-packages -no-user-package-db -package-db /home/sergey/projects/haskell/projects/sicp/scheme/.cabal-sandbox/x86_64-linux-ghc-7.8.3-packages.conf.d -package-db /tmp/dist/package.conf.inplace -package-id array-0.5.0.0-98aa445e59f3eb0c886795ff07406d84 -package-id base-4.7.0.1-e4b74d27ad8c8987c63abc42a80e7335 -package-id containers-0.5.5.1-23e2a2b94d6e452c773209f31d8672c5 -package-id mtl-2.1.3.1-8bcc0591131896cfc8761a93703d4c61 -package-id semigroups-0.15.3-90749480009265ad7841522aa7d8fc70 -package-id template-haskell-2.9.0.0-3afdeeadb657091939ef0bfa600dc614 -package-id text-1.1.1.3-19b1d34e1f78946a216fb2a95da8973b -package-id uu-parsinglib-2.8.1.1-2f10f272995445e8d7eae85e3789a3de -package-id vector-0.10.12.1-c5a0156e328bb6107669589785f49d3a -package-id void-0.6.1-db58af4741d5e0ed0ea75f96434b5878 -package-id wl-pprint-text-1.1.0.2-e332be893b2f20d19ec890651161d213 -XHaskell2010 Eval Eval.EvalM Eval.ExtractProgram Eval.Desugar ALaCarte ALaCarte.TH Parse Types SexpFrontend.Lexer SexpFrontend.LexTok SexpFrontend.Parser Utils -fhpc -hpcdir /tmp/dist/hpc/mix/scheme-0.1 -W -Wall -fwarn-monomorphism-restriction -fno-warn-name-shadowing -ferror-spans -rtsopts -W -Wall -fwarn-monomorphism-restriction -fno-warn-name-shadowing -ferror-spans -rtsopts
Linking...
/usr/bin/ar -r /tmp/dist/build/libHSscheme-0.1.a-15233/libHSscheme-0.1.a /tmp/dist/build/Eval.o /tmp/dist/build/Eval/EvalM.o /tmp/dist/build/Eval/ExtractProgram.o /tmp/dist/build/Eval/Desugar.o /tmp/dist/build/ALaCarte.o /tmp/dist/build/ALaCarte/TH.o /tmp/dist/build/Parse.o /tmp/dist/build/Types.o /tmp/dist/build/SexpFrontend/Lexer.o /tmp/dist/build/SexpFrontend/LexTok.o /tmp/dist/build/SexpFrontend/Parser.o /tmp/dist/build/Utils.o
/usr/bin/ar: creating /tmp/dist/build/libHSscheme-0.1.a-15233/libHSscheme-0.1.a
/usr/bin/strip /tmp/dist/build/libHSscheme-0.1.a-15233/libHSscheme-0.1.a --strip-unneeded
/usr/bin/ld -x -r -o /tmp/dist/build/HSscheme-0.1.o /tmp/dist/build/Eval.o /tmp/dist/build/Eval/EvalM.o /tmp/dist/build/Eval/ExtractProgram.o /tmp/dist/build/Eval/Desugar.o /tmp/dist/build/ALaCarte.o /tmp/dist/build/ALaCarte/TH.o /tmp/dist/build/Parse.o /tmp/dist/build/Types.o /tmp/dist/build/SexpFrontend/Lexer.o /tmp/dist/build/SexpFrontend/LexTok.o /tmp/dist/build/SexpFrontend/Parser.o /tmp/dist/build/Utils.o
/home/sergey/projects/haskell/ghc/local/bin/ghc -shared -dynamic -package-name scheme-0.1 -no-auto-link-packages -no-user-package-db -package-db /home/sergey/projects/haskell/projects/sicp/scheme/.cabal-sandbox/x86_64-linux-ghc-7.8.3-packages.conf.d -package-db /tmp/dist/package.conf.inplace -package-id array-0.5.0.0-98aa445e59f3eb0c886795ff07406d84 -package-id base-4.7.0.1-e4b74d27ad8c8987c63abc42a80e7335 -package-id containers-0.5.5.1-23e2a2b94d6e452c773209f31d8672c5 -package-id mtl-2.1.3.1-8bcc0591131896cfc8761a93703d4c61 -package-id semigroups-0.15.3-90749480009265ad7841522aa7d8fc70 -package-id template-haskell-2.9.0.0-3afdeeadb657091939ef0bfa600dc614 -package-id text-1.1.1.3-19b1d34e1f78946a216fb2a95da8973b -package-id uu-parsinglib-2.8.1.1-2f10f272995445e8d7eae85e3789a3de -package-id vector-0.10.12.1-c5a0156e328bb6107669589785f49d3a -package-id void-0.6.1-db58af4741d5e0ed0ea75f96434b5878 -package-id wl-pprint-text-1.1.0.2-e332be893b2f20d19ec890651161d213 /tmp/dist/build/Eval.dyn_o /tmp/dist/build/Eval/EvalM.dyn_o /tmp/dist/build/Eval/ExtractProgram.dyn_o /tmp/dist/build/Eval/Desugar.dyn_o /tmp/dist/build/ALaCarte.dyn_o /tmp/dist/build/ALaCarte/TH.dyn_o /tmp/dist/build/Parse.dyn_o /tmp/dist/build/Types.dyn_o /tmp/dist/build/SexpFrontend/Lexer.dyn_o /tmp/dist/build/SexpFrontend/LexTok.dyn_o /tmp/dist/build/SexpFrontend/Parser.dyn_o /tmp/dist/build/Utils.dyn_o -o /tmp/dist/build/libHSscheme-0.1-ghc7.8.3.so
In-place registering scheme-0.1...
/home/sergey/projects/haskell/ghc/local/bin/ghc-pkg update - --global --no-user-package-db '--package-db=/home/sergey/projects/haskell/projects/sicp/scheme/.cabal-sandbox/x86_64-linux-ghc-7.8.3-packages.conf.d' '--package-db=/tmp/dist/package.conf.inplace'
Preprocessing test suite 'scheme-test' for scheme-0.1...
Building test suite scheme-test...
creating /tmp/dist/build/scheme-test
creating /tmp/dist/build/scheme-test/scheme-test-tmp
/home/sergey/projects/haskell/ghc/local/bin/ghc --make -no-link -fbuilding-cabal-package -O2 -static -outputdir /tmp/dist/build/scheme-test/scheme-test-tmp -odir /tmp/dist/build/scheme-test/scheme-test-tmp -hidir /tmp/dist/build/scheme-test/scheme-test-tmp -stubdir /tmp/dist/build/scheme-test/scheme-test-tmp -i -i/tmp/dist/build/scheme-test/scheme-test-tmp -itest -i/tmp/dist/build/autogen -I/tmp/dist/build/autogen -I/tmp/dist/build/scheme-test/scheme-test-tmp -optP-include -optP/tmp/dist/build/autogen/cabal_macros.h -hide-all-packages -no-user-package-db -package-db /home/sergey/projects/haskell/projects/sicp/scheme/.cabal-sandbox/x86_64-linux-ghc-7.8.3-packages.conf.d -package-db /tmp/dist/package.conf.inplace -package-id array-0.5.0.0-98aa445e59f3eb0c886795ff07406d84 -package-id base-4.7.0.1-e4b74d27ad8c8987c63abc42a80e7335 -package-id containers-0.5.5.1-23e2a2b94d6e452c773209f31d8672c5 -package-id mtl-2.1.3.1-8bcc0591131896cfc8761a93703d4c61 -package-id semigroups-0.15.3-90749480009265ad7841522aa7d8fc70 -package-id tasty-0.10.0.2-e97431e21992083bf0a2dafc10227580 -package-id tasty-hunit-0.9.0.1-5e5151e20820df447ef956bc3c845181 -package-id template-haskell-2.9.0.0-3afdeeadb657091939ef0bfa600dc614 -package-id text-1.1.1.3-19b1d34e1f78946a216fb2a95da8973b -package-id uu-parsinglib-2.8.1.1-2f10f272995445e8d7eae85e3789a3de -package-id vector-0.10.12.1-c5a0156e328bb6107669589785f49d3a -package-id void-0.6.1-db58af4741d5e0ed0ea75f96434b5878 -package-id wl-pprint-text-1.1.0.2-e332be893b2f20d19ec890651161d213 -package-id scheme-0.1-inplace -XHaskell2010 test/Main.hs -fhpc -hpcdir /tmp/dist/hpc/mix/scheme-test
Linking...
/home/sergey/projects/haskell/ghc/local/bin/ghc --make -fbuilding-cabal-package -O2 -static -outputdir /tmp/dist/build/scheme-test/scheme-test-tmp -odir /tmp/dist/build/scheme-test/scheme-test-tmp -hidir /tmp/dist/build/scheme-test/scheme-test-tmp -stubdir /tmp/dist/build/scheme-test/scheme-test-tmp -i -i/tmp/dist/build/scheme-test/scheme-test-tmp -itest -i/tmp/dist/build/autogen -I/tmp/dist/build/autogen -I/tmp/dist/build/scheme-test/scheme-test-tmp -optP-include -optP/tmp/dist/build/autogen/cabal_macros.h -hide-all-packages -no-user-package-db -package-db /home/sergey/projects/haskell/projects/sicp/scheme/.cabal-sandbox/x86_64-linux-ghc-7.8.3-packages.conf.d -package-db /tmp/dist/package.conf.inplace -package-id array-0.5.0.0-98aa445e59f3eb0c886795ff07406d84 -package-id base-4.7.0.1-e4b74d27ad8c8987c63abc42a80e7335 -package-id containers-0.5.5.1-23e2a2b94d6e452c773209f31d8672c5 -package-id mtl-2.1.3.1-8bcc0591131896cfc8761a93703d4c61 -package-id semigroups-0.15.3-90749480009265ad7841522aa7d8fc70 -package-id tasty-0.10.0.2-e97431e21992083bf0a2dafc10227580 -package-id tasty-hunit-0.9.0.1-5e5151e20820df447ef956bc3c845181 -package-id template-haskell-2.9.0.0-3afdeeadb657091939ef0bfa600dc614 -package-id text-1.1.1.3-19b1d34e1f78946a216fb2a95da8973b -package-id uu-parsinglib-2.8.1.1-2f10f272995445e8d7eae85e3789a3de -package-id vector-0.10.12.1-c5a0156e328bb6107669589785f49d3a -package-id void-0.6.1-db58af4741d5e0ed0ea75f96434b5878 -package-id wl-pprint-text-1.1.0.2-e332be893b2f20d19ec890651161d213 -package-id scheme-0.1-inplace -XHaskell2010 test/Main.hs -o /tmp/dist/build/scheme-test/scheme-test -fhpc -hpcdir /tmp/dist/hpc/mix/scheme-test
Running 1 test suites...
Test suite scheme-test: RUNNING...
/tmp/dist/build/scheme-test/scheme-test
Tests
  Parser tests
    Parse integer:                                         OK
    Parse nil:                                             OK
    Parse double:                                          OK
    Parse symbol:                                          OK
    Parse cons pair:                                       OK
    Parse sample sexp:                                     OK
    Parse empty sexp:                                      OK
    Parse tricky sexp that tries to look like dotted list: OK
    Parse singleton int sexp:                              OK
    Parse two-element int sexp:                            OK
    Parse two-element int dot-list sexp:                   OK
    Parse singleton vector:                                OK
    Parse sample vector:                                   OK
  Extractor tests
    Extract atom:                                          OK
    Extract let:                                           OK
    Extract and:                                           OK
    Extract and error #1:                                  OK
    Extract and error #2:                                  OK
    Extract and error #3:                                  OK
    Extract set!:                                          OK
    Extract set! error #1:                                 OK
    Extract set! error #2:                                 OK
    Extract set! error #3:                                 OK
  Eval tests
    self-evaluating integers:                              OK
    self-evaluating nil:                                   OK
    assignment:                                            OK
    define:                                                OK
    define and assignment:                                 OK
    lambda application:                                    OK
    if #1:                                                 OK
    if #2:                                                 OK
    lambda counter with state:                             OK
    or short-circuit:                                      OK
    and short-circuit:                                     OK
    sum #1:                                                OK
    sum #2:                                                OK
    product #1:                                            OK
    product #2:                                            OK
    difference:                                            OK
    difference err #1:                                     OK
    division:                                              OK
    division err #1:                                       OK
    division err #2:                                       OK
    division err #3:                                       OK
    null? #1:                                              OK
    null? #2:                                              OK
    cons #1:                                               OK
    cons #2:                                               OK
    cons #3:                                               OK
    cons err #1:                                           OK
    cons err #2:                                           OK
    cons err #3:                                           OK
    car #1:                                                OK
    car err #1:                                            OK
    car err #2:                                            OK
    car err #3:                                            OK
    cdr #1:                                                OK
    cdr err #1:                                            OK
    cdr err #2:                                            OK
    cdr err #3:                                            OK

All 60 tests passed (0.01s)
Test suite scheme-test: PASS
Test suite logged to: /tmp/dist/test/scheme-0.1-scheme-test.log
/home/sergey/projects/haskell/ghc/local/bin/hpc markup /tmp/dist/hpc/tix/scheme-test/scheme-test.tix --destdir=/tmp/dist/hpc/html/scheme-test --hpcdir=/tmp/dist/hpc/mix/scheme-test --hpcdir=/tmp/dist/hpc/mix/scheme-0.1 --exclude=Main
hpc: can not find scheme-0.1/SexpFrontend.LexTok in ["./.hpc",".//tmp/dist/hpc/mix/scheme-0.1",".//tmp/dist/hpc/mix/scheme-test"]
```

Second (abbreviated) log illustrating the problem with `--hpc-options='--srcdir="" --srcdir="."`:

```
cd /home/sergey/projects/haskell/projects/sicp/scheme/ && \
cabal configure --enable-library-coverage --disable-library-profiling --disable-executable-profiling --disable-split-objs --hpc-options='--srcdir="" --srcdir="."' --builddir /tmp/dist --enable-tests && \
cabal build -v1 --builddir /tmp/dist --ghc-options="-j4 -ferror-spans" && \
cabal test -v2 --builddir /tmp/dist --show-details=always
Resolving dependencies...
Configuring scheme-0.1...
Building scheme-0.1...
Preprocessing library scheme-0.1...
In-place registering scheme-0.1...
Preprocessing test suite 'scheme-test' for scheme-0.1...
Using a sandbox located at
/home/sergey/projects/haskell/projects/sicp/scheme/.cabal-sandbox
Reading available packages...
Reading installed packages...
/home/sergey/projects/haskell/ghc/local/bin/ghc-pkg dump '--package-db=/home/sergey/projects/haskell/projects/sicp/scheme/.cabal-sandbox/x86_64-linux-ghc-7.8.3-packages.conf.d' -v0
/home/sergey/projects/haskell/ghc/local/bin/ghc --print-libdir
Found no modified add-source deps.
Component build order: library, test suite 'scheme-test'
creating /tmp/dist/build
creating /tmp/dist/build/autogen
Building scheme-0.1...
Preprocessing library scheme-0.1...
Building library...
...
In-place registering scheme-0.1...
...
Preprocessing test suite 'scheme-test' for scheme-0.1...
Building test suite scheme-test...
...
Running 1 test suites...
Test suite scheme-test: RUNNING...
/tmp/dist/build/scheme-test/scheme-test
Tests
...

All 60 tests passed (0.01s)
Test suite scheme-test: PASS
Test suite logged to: /tmp/dist/test/scheme-0.1-scheme-test.log
/home/sergey/projects/haskell/ghc/local/bin/hpc markup /tmp/dist/hpc/tix/scheme-test/scheme-test.tix --destdir=/tmp/dist/hpc/html/scheme-test --hpcdir=/tmp/dist/hpc/mix/scheme-test --hpcdir=/tmp/dist/hpc/mix/scheme-0.1 --exclude=Main --srcdir= --srcdir=.
Writing: scheme-0.1/SexpFrontend.LexTok.hs.html
Writing: scheme-0.1/SexpFrontend.Lexer.hs.html
Writing: scheme-0.1/Utils.hs.html
Writing: scheme-0.1/SexpFrontend.Parser.hs.html
Writing: scheme-0.1/Types.hs.html
Writing: scheme-0.1/ALaCarte.hs.html
Writing: scheme-0.1/Eval.Desugar.hs.html
Writing: scheme-0.1/Eval.ExtractProgram.hs.html
Writing: scheme-0.1/Eval.EvalM.hs.html
Writing: scheme-0.1/Eval.hs.html
Writing: hpc_index.html
Writing: hpc_index_fun.html
Writing: hpc_index_alt.html
Writing: hpc_index_exp.html
Test coverage report written to /tmp/dist/hpc/html/scheme-test/hpc_index.html
1 of 1 test suites (1 of 1 test cases) passed.
/home/sergey/projects/haskell/ghc/local/bin/hpc sum --union /tmp/dist/hpc/tix/scheme-test/scheme-test.tix --output=/tmp/dist/hpc/tix/scheme-0.1/scheme-0.1.tix --exclude=Main --srcdir= --srcdir=.
hpc failed:
  unrecognized option `--srcdir='
  unrecognized option `--srcdir=.'


Usage: hpc sum [OPTION] .. <TIX_FILE> [<TIX_FILE> [<TIX_FILE> ..]]
Sum multiple .tix files in a single .tix file


Options:

    --exclude=[PACKAGE:][MODULE]  exclude MODULE and/or PACKAGE
    --include=[PACKAGE:][MODULE]  include MODULE and/or PACKAGE
    --output=FILE                 output FILE
    --union                       use the union of the module namespace (default is intersection)


```

**Comment by thomie** [comment]
Filed as https://ghc.haskell.org/trac/ghc/ticket/10138.

**Comment by thomie** [comment]
This is fixed in GHC 7.12.
